src/sdk/namespace_fs.js

```diff
@@ -960,34 +996,39 @@ class NamespaceFS {
         let file_path;
         try {
             await this._load_bucket(params, fs_context);
-            file_path = await this._find_version_path(fs_context, params);
-            await this._check_path_in_bucket_boundaries(fs_context, file_path);
-
-            // NOTE: don't move this code after the open
-            // this can lead to ENOENT failures due to file not exists when content size is 0
-            // if entry is a directory object and its content size = 0 - return empty response
-            const is_dir_content = this._is_directory_content(file_path, params.key);
-            if (is_dir_content) {
+            let retries = (this._is_versioning_disabled() || this._is_versioning_suspended()) ? config.NSFS_RENAME_RETRIES : 0;
+            const is_gpfs = native_fs_utils._is_gpfs(fs_context);
+            let stat;
+            for (;;) {
                 try {
-                    const md_path = this._get_file_md_path(params);
-                    const dir_stat = await nb_native().fs.stat(fs_context, md_path);
-                    if (dir_stat && dir_stat.xattr[XATTR_DIR_CONTENT] === '0') return null;
+                    file_path = await this._find_version_path(fs_context, params);
+                    await this._check_path_in_bucket_boundaries(fs_context, file_path);
+
+                    // NOTE: don't move this code after the open
+                    // this can lead to ENOENT failures due to file not exists when content size is 0
+                    // if entry is a directory object and its content size = 0 - return empty response
+                    if (await this._is_empty_directory_content(file_path, fs_context, params)) return null;
+
+                    file = await nb_native().fs.open(
+                        fs_context,
+                        file_path,
+                        config.NSFS_OPEN_READ_MODE,
+                        native_fs_utils.get_umasked_mode(config.BASE_MODE_FILE),
+                    );
+                    stat = await file.stat(fs_context);
+                    if (params.version_id && !this._is_versioning_disabled() && this._get_version_id_by_xattr(stat) !== params.version_id) {
+                        retries -= 1;
+                        if (retries <= 0) throw new Error('to many retries - mismatch version_id');
+                        dbg.warn('NamespaceFS.read_object_stream mismatch version_id', params.version_id, this._get_version_id_by_xattr(stat));
+                        continue;
+                    }
+                    break;
                 } catch (err) {
-                    //failed to get object
-                    new NoobaaEvent(NoobaaEvent.OBJECT_GET_FAILED).create_event(params.key,
```

Comment:
can you bring back this event when throwing an error?

Reply:
this even still exists I just moved it in _is_empty_directory_content with the rest of the code

Reply:
I saw it, but it should be here, we should emit this event not only on the catch of _is_empty_directory_content () and when is_dir_content is true